drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2149 @@
+	}
+
+	return err;
+}
+#endif
+
+#ifdef CONFIG_I3C_STM32_DMA
+static int i3c_stm32_dma_stream_config(const struct device *dev,
+				       struct i3c_stm32_dma_stream *dma_stream, uint64_t src_addr,
+				       uint64_t dst_addr)
+{
+	if (dma_stream->dma_dev != NULL) {
+		if (!device_is_ready(dma_stream->dma_dev)) {
+			return -ENODEV;
+		}
+	}
+
+	memset(&dma_stream->blk_cfg, 0, sizeof(dma_stream->blk_cfg));
+
+	dma_stream->blk_cfg.source_address = src_addr;
+
+	dma_stream->blk_cfg.dest_address = dst_addr;
+
+	if (dma_stream->src_addr_increment) {
+		dma_stream->blk_cfg.source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	} else {
+		dma_stream->blk_cfg.source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+	}
+
+	if (dma_stream->dst_addr_increment) {
+		dma_stream->blk_cfg.dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	} else {
+		dma_stream->blk_cfg.dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+	}
+
+	dma_stream->blk_cfg.source_reload_en = 0;
+	dma_stream->blk_cfg.dest_reload_en = 0;
+	dma_stream->blk_cfg.fifo_mode_control = dma_stream->fifo_threshold;
+
+	dma_stream->dma_cfg.head_block = &dma_stream->blk_cfg;
+	dma_stream->dma_cfg.user_data = (void *)dev;
+
+	return 0;
+}
+
+/* Initializes the I3C DMA */
+static int i3c_stm32_init_dma(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	int err;
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+
+	/*Configure DMA RX */
+	err = i3c_stm32_dma_stream_config(
+		dev, &data->dma_rx, LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_DATA_RECEIVE_BYTE),
+		0);
+	if (err != 0) {
+		return err;
+	}
+
+	/*Configure DMA RS */
+	err = i3c_stm32_dma_stream_config(dev, &data->dma_rs,
+					  LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_STATUS), 0);
+	if (err != 0) {
+		return err;
+	}
+
+	/*Configure DMA TX */
+	err = i3c_stm32_dma_stream_config(
+		dev, &data->dma_tx, 0,
+		LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_DATA_TRANSMIT_BYTE));
+	if (err != 0) {
+		return err;
+	}
+
+	/*Configure DMA TC */
+	err = i3c_stm32_dma_stream_config(dev, &data->dma_tc, 0,
+					  LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_CONTROL));
+	if (err != 0) {
+		return err;
+	}
+
+	return err;
+}
+#endif
+
+static void i3c_stm32_controller_init(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+
+#ifdef CONFIG_I3C_USE_IBI
+	LL_I3C_EnableHJAck(i3c);
+#endif
+
+	/* Configure FIFO */
+	LL_I3C_SetRxFIFOThreshold(i3c, LL_I3C_RXFIFO_THRESHOLD_1_4);
+	LL_I3C_SetTxFIFOThreshold(i3c, LL_I3C_TXFIFO_THRESHOLD_1_4);
+	LL_I3C_EnableControlFIFO(i3c);
+	LL_I3C_EnableStatusFIFO(i3c);
+
+	/* I3C Initialization */
+	LL_I3C_SetMode(i3c, LL_I3C_MODE_CONTROLLER);
+	LL_I3C_SetStallTime(i3c, 0x00);
+	LL_I3C_DisableStallACK(i3c);
+	LL_I3C_DisableStallParityCCC(i3c);
+	LL_I3C_DisableStallParityData(i3c);
+	LL_I3C_DisableStallTbit(i3c);
+	LL_I3C_DisableHighKeeperSDA(i3c);
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
```

Comment:
Thinking more about this, This is related to the kernel clock input and meeting the required t_hd_pp time, and should be within `i3c_stm32_config_ctrl_bus_char`... but I would expect some math behind it to determine if it should be 0.5 or 1.5.

Reply:
According to the reference manual for the H5 board, the value of the minimum data hold time can be calculated as follows:

- $t_{HDPP_{min}} = min(t_{CR}, t_{CF}) + 3ns$

We could not find how the $t_{CR}$ and $t_{CF}$ values can be calculated so we chose 1.5 data hold time as default for now to avoid any issues that might occur when choosing 0.5.

If you have any suggestion on how we can dynamically set these values according to the $t_{CR}$, $t_{CF}$ and the kernel clock, we can modify the code accordingly.